File: ios/Classes/Managers/MethodChannelCardMerchantManager.swift
//
//  MethodChannelMerchantManager.swift
//  Runner
//
//  Created by Ray Chow on 26/4/2022.
//

import Foundation
import Flutter
import RxSwift

class MethodChannelCardMerchantManager: BaseMethodChannelManager {
    static let shared = MethodChannelCardMerchantManager()
    let disposeBag = DisposeBag()
    
    func handleMethod(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let method = MethodChannelCardMerchantMethodKey(rawValue:  call.method) else {
            result(FlutterMethodNotImplemented)
            return
        }
        
        guard let arguments = call.arguments as? Dictionary<String, Any>,
                let data = arguments[MethodChannelRequestKey.data.rawValue] as? Dictionary<String, Any>
         else {
            result(FlutterError(code: "-1", message: "Cannot find object with key - data", details: nil))
            return
        }
        
        switch method {
            //MPGS
        case .tokeniseWithMpgs:
            MpgsMerchantManager.shared.tokeniseWithMpgs(data: data).subscribe(with: self) { this, mpgsTokeniseResult in
                result(this.mapToMethodChannelResultDataDictionary(
                    data: mpgsTokeniseResult.convertToDict()))
            } onFailure: { this, error in
                result(FlutterError(code: "-1", message: error.localizedDescription, details: nil))
            }.disposed(by: disposeBag)
            break
        case .showMpgs3ds:
            MpgsMerchantManager.shared.showMpgs3ds(data: data).take(1).subscribe(with: self) { this, mpgs3dsResult in
                result(this.mapToMethodChannelResultDataDictionary(data: mpgs3dsResult.convertToDict()))
            } onError: { this, error in
                result(FlutterError(code: "-1", message: error.localizedDescription, details: nil))
            }.disposed(by: disposeBag)
            break
        default:
            result(FlutterMethodNotImplemented)
            break
        }
    }
}

File: android/src/main/kotlin/com/keychainpay/keychainpay_mpgs/manager/method_channel/card_merchant/MethodChannelCardMerchantManager.kt
package com.keychainpay.keychainpay_mpgs.manager.method_channel.card_merchant

import android.content.Intent
import androidx.activity.result.ActivityResultLauncher
import androidx.annotation.NonNull
import com.keychainpay.keychainpay_mpgs.enums.method_channel.card_merchant.common.method_key.MethodChannelCardMerchantMethodKey
import com.keychainpay.keychainpay_mpgs.enums.method_channel.common.MethodChannelRequestKey
import com.keychainpay.keychainpay_mpgs.extension.disposedBy
import com.keychainpay.keychainpay_mpgs.extension.runInBackground
import com.keychainpay.keychainpay_mpgs.manager.card_merchant.mpgs.MpgsMerchantManager
import com.keychainpay.keychainpay_mpgs.manager.method_channel.base.BaseMethodChannelManager
import io.flutter.embedding.android.FlutterFragmentActivity
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import io.reactivex.disposables.CompositeDisposable

class MethodChannelCardMerchantManager: BaseMethodChannelManager() {
    val disposeBag = CompositeDisposable()
    companion object {
        private var instance: MethodChannelCardMerchantManager? = null

        fun getInstance(): MethodChannelCardMerchantManager {
            if (instance == null) {
                instance = MethodChannelCardMerchantManager()
            }
            return instance!!
        }
    }

    fun handleMethod(@NonNull call: MethodCall, @NonNull result: Result, activity: FlutterFragmentActivity, mpgs3DSecureLauncher: ActivityResultLauncher<Intent>) {
        val method = MethodChannelCardMerchantMethodKey.values()
            .firstOrNull { methodKey -> methodKey.key == call.method }?: run {
                result.notImplemented()
                return
            }
        val data: HashMap<String, Any> = call.argument(MethodChannelRequestKey.DATA.key)?: run {
            result.error("-1", "Cannot find object with key - data", null)
            return
        }

        when (method) {
            //MPGS
            MethodChannelCardMerchantMethodKey.TOKENISE_WITH_MPGS -> {
                MpgsMerchantManager.getInstance().tokeniseCreditCardWithMpgs(data)
                    .runInBackground()
                    .subscribe({
                        result.success(mapToMethodChannelDataMap(it.convertToHashMap()))
                    }, {
                        result.error("-1", it.localizedMessage, null)
                    }).disposedBy(disposeBag)
            }
            MethodChannelCardMerchantMethodKey.SHOW_MPGS_3DS -> {
                MpgsMerchantManager.getInstance().showMpgs3ds(data, mpgs3DSecureLauncher, activity)
                    .subscribe({
                        result.success(mapToMethodChannelDataMap(it.convertToHashMap()))
                    },{
                        result.error("-1", it.localizedMessage, null)
                    })
                    .disposedBy(disposeBag)
            }
            else -> result.notImplemented()
        }
    }
}